base_layer/core/src/transactions/transaction_components/side_chain/contract_constitution.rs

```diff
@@ -0,0 +1,339 @@
+    fn consensus_decode<R: Read>(reader: &mut R) -> Result<Self, Error> {
+        Ok(Self {
+            period_expiry: u64::consensus_decode(reader)?,
+            minimum_quorum_required: u32::consensus_decode(reader)?,
+        })
+    }
+}
+
+#[derive(Debug, Clone, Hash, PartialEq, Deserialize, Serialize, Eq)]
+pub struct CheckpointParameters {
+    /// The minimum number of votes (signatures) required on each checkpoint.
+    pub quorum_minimum: u32,
+    /// If this number of blocks have passed without a checkpoint, the contract becomes abandoned.
+    pub abandoned_interval: u64,
+}
+
+impl ConsensusEncoding for CheckpointParameters {
+    fn consensus_encode<W: Write>(&self, writer: &mut W) -> Result<(), Error> {
+        self.quorum_minimum.consensus_encode(writer)?;
+        self.abandoned_interval.consensus_encode(writer)?;
+        Ok(())
+    }
+}
+
+impl ConsensusEncodingSized for CheckpointParameters {}
+
+impl ConsensusDecoding for CheckpointParameters {
+    fn consensus_decode<R: Read>(reader: &mut R) -> Result<Self, Error> {
+        Ok(Self {
+            quorum_minimum: u32::consensus_decode(reader)?,
+            abandoned_interval: u64::consensus_decode(reader)?,
+        })
+    }
+}
+
+#[derive(Debug, Clone, Hash, PartialEq, Deserialize, Serialize, Eq)]
+pub struct ConstitutionChangeRules {
+    /// Bitflag that indicates the constitution changes that are permitted.
+    pub change_flags: ConstitutionChangeFlags,
+    /// The rules for changes to the validator committee.
+    pub committee_change_rules: CommitteeChangeRules,
+    /// Requirements for amendments to the contract constitution. If None, then the `ContractConstitution` cannot be
+    /// changed.
+    pub requirements_for_constitution_change: Option<RequirementsForConstitutionChange>,
+}
+
+impl ConsensusEncoding for ConstitutionChangeRules {
+    fn consensus_encode<W: Write>(&self, writer: &mut W) -> Result<(), Error> {
+        self.change_flags.consensus_encode(writer)?;
+        self.committee_change_rules.consensus_encode(writer)?;
+        self.requirements_for_constitution_change.consensus_encode(writer)?;
+        Ok(())
+    }
+}
+
+impl ConsensusEncodingSized for ConstitutionChangeRules {}
+
+impl ConsensusDecoding for ConstitutionChangeRules {
+    fn consensus_decode<R: Read>(reader: &mut R) -> Result<Self, Error> {
+        Ok(Self {
+            change_flags: ConstitutionChangeFlags::consensus_decode(reader)?,
+            committee_change_rules: CommitteeChangeRules::consensus_decode(reader)?,
+            requirements_for_constitution_change: ConsensusDecoding::consensus_decode(reader)?,
+        })
+    }
+}
+
+#[derive(Debug, Clone, Hash, PartialEq, Deserialize, Serialize, Eq)]
+pub struct CommitteeChangeRules {
+    /// The maximum number of committee members that can be added.
+    pub max_new_committee_members: u32,
+    /// The maximum number of committee members that can be evicted.
+    pub max_evicted_committee_members: u32,
```

Comment:
Perhaps more options are better, but I don't see any reason why we would limit these two fields

Reply:
Yeah wasn't sure what settings we want to add to this. Should only have settings that are useful since each setting translates into more validation complexity and impact performance. Suggestions welcome. Could also leave this out for now.